### ISSUE TITLE
Estimate the required space

### ISSUE THREAD

**Issue by davidmef** [stevesoltys/backup#363]
I made a backup to an SD card and it failed for lack of space. It worked when I excluded .apk from the backup. Could you provide an estimate of the required space before the backup? I guess you don't because you would have to do the actual backup to get result, but an estimate with margins ("between 1.2 and 3.4 GB") would be better than nothing, I guess.

**Comment by grote** [comment]
We might be able to get the size required for APK backups, but most likely not the backup data because like you said, there's no API for that, we'd need to try a backup to find out.

**Comment by stevleibelt** [comment]
@davidmef  do you feel like @grote   was able to answer your question?

If so, please close this ticket.

**Comment by nettnikl** [comment]
@grote Could you provide detail on 'try to backup'? If you think it can be done by 'scanning the file tree, similar to what is the first step of backing up', that seems to be what was requested here, a way to execute this very first part of the process. The margin could then not be estimated, but would only be the size of files written in the time between the estimation and the actual backup. And the implementation effort would be relatively limited (i assume).

Please do not close the ticket until we found a solution, it does'nt make sense in my view to close tickets that are not solved, no one here is paid by the closed ticket.

**Comment by grote** [comment]
> Could you provide detail on 'try to backup'? If you think it can be done by 'scanning the file tree, similar to what is the first step of backing up',

No, what I mean is that we have to do an actual backup first to know what size it will have in the end as Android is not telling us before. For file and APK backup, we can find out before.

**Comment by nettnikl** [comment]
Okay, looking at the different types of data we can backup:
- APKs
- App data <- this is what we cannot calculate properly, right?
- Plain-old file backup

Sorry, i'm not familiar with the details, but is there no API for checking this? Or can this be done by just checking how big the app data is in the filesystem?

**Comment by dnut** [comment]
There must be a solution, since seedvault is able to access all the data when running a backup.

At the very least, couldn’t seedvault support a dry run that executes the normal backup pipeline, but swaps the writer for a no-op/counting sink? That way it could determine exactly how many bytes would actually be written without producing any effects.

**Comment by grote** [comment]
You could do that, but it would require quite complex code changes and the UX wouldn't be great. We already have detection and warnings for when the backup location runs out of space, so why not simply try a backup where you plan to store it and see if it works?
